src/geode/io/mesh/vtp_input.cpp

```diff
@@ -0,0 +1,424 @@
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+#include <geode/io/mesh/private/vtp_input.h>
+
+#include <absl/strings/escaping.h>
+#include <absl/strings/numbers.h>
+#include <bitset>
+#include <fstream>
+#include <iostream>
+#include <locale>
+#include <pugixml.hpp>
+#include <zlib-ng.h>
+
+#include <geode/geometry/point.h>
+
+#include <geode/mesh/builder/polygonal_surface_builder.h>
+#include <geode/mesh/core/geode_polygonal_surface.h>
+
+namespace
+{
+    absl::string_view remove_spaces( absl::string_view in )
+    {
+        auto left = in.begin();
+        while( isspace( *left ) )
+        {
+            left++;
+        }
+        for( ;; ++left )
+        {
+            if( left == in.end() )
+                return absl::string_view();
+            if( !isspace( *left ) )
+                break;
+        }
+        auto right = in.end() - 1;
+        while( isspace( *right ) && right > left )
+        {
+            right--;
+        }
+        return absl::string_view( left, std::distance( left, right ) + 1 );
+    }
+
+    class VTPInputImpl
+    {
+    public:
+        VTPInputImpl( absl::string_view filename,
+            geode::PolygonalSurface3D& polygonal_surface )
+            : file_( filename.data() ),
+              surface_( polygonal_surface ),
+              surface_builder_{ geode::PolygonalSurfaceBuilder3D::create(
+                  polygonal_surface ) }
+        {
+            OPENGEODE_EXCEPTION( file_.good(),
+                "[VTPInput] Error while opening file: ", filename );
+            const auto ok = document_.load_file( filename.data() );
+            OPENGEODE_EXCEPTION(
+                ok, "[VTPInput] Error while parsing file: ", filename );
+            root_ = document_.child( "VTKFile" );
+        }
+
+        void read_file()
+        {
+            read_root_attributes();
+            for( const auto& polydata : root_.children( "PolyData" ) )
+            {
+                read_polydata( polydata );
+            }
+        }
+
+    private:
+        void read_root_attributes()
+        {
+            OPENGEODE_EXCEPTION(
+                strcmp( root_.attribute( "type" ).value(), "PolyData" ) == 0,
+                "[VTPInput::read_root_attributes] VTK File type should be "
+                "PolyData" );
+            if( strcmp( root_.attribute( "byte_order" ).value(), "BigEndian" )
```

Comment:
Is this better https://github.com/abseil/abseil-cpp/blob/master/absl/strings/match.h